### PR TITLE
Be 48: Trips and Memberships

### PR DESCRIPTION
**Pull request by hill-ol**
# Description

This PR implements the Trips and Memberships feature, enabling users to create trips and manage trip membership with admin/member roles.

## Summary
Added complete CRUD operations for trips and membership management. Users can create trips, join trips as members, and be promoted/demoted to admin roles. Each membership tracks individual budget preferences.

## High-level Approach
- **Database Layer**: Created `trips` and `memberships` tables with composite primary key (user_id, trip_id) for memberships
- **Repository Layer**: Implemented repository pattern for data access with Bun ORM
- **Service Layer**: Added business logic for trip validation, membership management, and role-based operations (promote/demote)
- **Controller Layer**: Built RESTful endpoints with JWT authentication and request validation
- **Testing**: Comprehensive integration tests covering full lifecycle including edge cases

## Key Design Decisions
- Used composite primary key for memberships (user_id, trip_id) rather than separate ID
- Implemented idempotent deletes (DELETE returns 204 even if resource already deleted)
- Separate endpoints for promote/demote operations rather than generic update
- Budget validation enforced at service layer (min >= 0, max >= min)

## API Endpoints Added
**Trips:**
- `POST /api/v1/trips` - Create trip
- `GET /api/v1/trips` - Get all trips
- `GET /api/v1/trips/:tripID` - Get specific trip
- `PATCH /api/v1/trips/:tripID` - Update trip
- `DELETE /api/v1/trips/:tripID` - Delete trip

**Memberships:**
- `POST /api/v1/memberships` - Add member to trip
- `GET /api/v1/trips/:tripID/members` - Get trip members
- `GET /api/v1/trips/:tripID/members/:userID` - Get specific membership
- `PATCH /api/v1/trips/:tripID/members/:userID` - Update membership
- `DELETE /api/v1/trips/:tripID/members/:userID` - Remove member
- `POST /api/v1/trips/:tripID/members/:userID/promote` - Promote to admin
- `POST /api/v1/trips/:tripID/members/:userID/demote` - Demote to member
- `GET /api/v1/users/:userID/trips` - Get user's trips

## How has this been tested?

Added comprehensive integration tests:
- `TestTripLifecycle` - 9 test cases covering full CRUD lifecycle
- `TestMembershipLifecycle` - 11 test cases covering membership operations, role management, and edge cases

All tests pass. Test coverage includes:
- Happy path scenarios (create, read, update, delete)
- Error cases (404 for deleted resources, validation errors)
- Edge cases (idempotent deletes, duplicate prevention)
- Role-based operations (promote/demote admins)

Frontend types and React Query hooks generated via `bun kubb`.

## Migration Note
Database migrations are included in this PR for review but will be deployed separately. A follow-up PR will be created specifically for applying migrations to production after this feature PR is approved.

## Checklist
* [x] I have self-reviewed my code for readability, maintainability, performance, and added comments/documentation where necessary.
* [x] New and existing tests pass locally with my changes.
* [x] I have followed the project's coding standards and best practices.

### CODE REVIEW

File: backend/internal/migrations/20260123213543_create_membership.sql [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+    is_admin BOOLEAN NOT NULL,
+    created_at TIMESTAMP WITH TIME ZONE DEFAULT now(),
+    updated_at TIMESTAMP WITH TIME ZONE DEFAULT now(),
+    budget_min INTEGER NOT NULL DEFAULT 0 CHECK (budget_min >= 0),
+    budget_max INTEGER NOT NULL DEFAULT 0 CHECK (budget_max >= 0 AND budget_max >= budget_min),
+    availability JSONB NULL,
+    PRIMARY KEY (user_id, trip_id),
+    FOREIGN KEY (user_id) REFERENCES users(id) ON DELETE CASCADE,
+    FOREIGN KEY (trip_id) REFERENCES trips(id) ON DELETE CASCADE
+);
```
[the remark this answers is not in the document]
Comment (in-mai-space) [reply backend/internal/migrations/20260123213543_create_membership.sql]: this is a good point, please add index!

File: backend/internal/repository/membership_repository.go [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+	err = r.db.NewSelect().
+		Model(updatedMembership).
+		Where("user_id = ? AND trip_id = ?", userID, tripID).
+		Scan(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	return updatedMembership, nil
+}
```
[the remark this answers is not in the document]
Comment (in-mai-space) [reply backend/internal/repository/membership_repository.go]: yes, instead of doing update then do a select, it's two calls to db, so we can do an update and then returning, which will return a new updated row

File: backend/internal/services/membership_service.go [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+	if req.BudgetMin < 0 {
+		return nil, errors.New("budget minimum cannot be negative")
+	}
+
+	if req.BudgetMax < req.BudgetMin {
+		return nil, errors.New("budget maximum must be greater than or equal to minimum")
+	}
+
+	return s.Membership.Update(ctx, userID, tripID, &req)
+}
```
[the remark this answers is not in the document]
Comment (in-mai-space) [reply backend/internal/services/membership_service.go]: yeh good point to take a look here!

**Review by in-mai-space — CHANGES_REQUESTED** [review]
insane work guys, some comments here and there but overall looks good!

File: backend/internal/models/trips.go
Comment (aahiltn) [inline backend/internal/models/trips.go]: ah these should be pointers and not required since its updating the row! no worries just a quick fix

File: backend/internal/repository/trip_repository.go [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+// Delete removes a trip (idempotent)
+func (r *tripRepository) Delete(ctx context.Context, id uuid.UUID) error {
+	_, err := r.db.NewDelete().
+		Model((*models.Trip)(nil)).
+		Where("id = ?", id).
+		Exec(ctx)
+	
+	// Idempotent - don't error if already deleted
+	return err
+}
```
Comment (aahiltn) [inline backend/internal/repository/trip_repository.go]: eof :)

File: backend/internal/server/routers/membership_routes.go [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+	"toggo/internal/types"
+
+	"github.com/gofiber/fiber/v2"
+)
+
+func MembershipRoutes(apiGroup fiber.Router, routeParams types.RouteParams) fiber.Router {
+	membershipService := services.NewMembershipService(routeParams.ServiceParams.Repository)
+	membershipController := controllers.NewMembershipController(membershipService, routeParams.Validator)
+
+	// /api/v1/memberships
```
Comment (aahiltn) [inline backend/internal/server/routers/membership_routes.go]: pls make this match the endpoint underneath :)
you can post with the trips/:tripId

**Review by in-mai-space — APPROVED** [review]
